runtime/engine.cpp

```diff
@@ -353,20 +353,39 @@ void api::Engine::abort(const char *reason) { ::abort(CONTEXT, reason); }
 bool api::Engine::eval_toplevel(const char *path, MutableHandleValue result) {
   JSContext *cx = CONTEXT;
   RootedValue ns(cx);
-  if (!scriptLoader->load_top_level_script(path, &ns)) {
+  RootedValue tla_promise(cx);
+  if (!scriptLoader->load_top_level_script(path, &ns, &tla_promise)) {
     return false;
   }
 
+  // Drive TLA as far as possible. Rejections during pre-initialization are
+  // treated as top-level exceptions. TLA may remain unresolved, in which case
+  // it will continue tasks at runtime. Rejections after pre-intialization remain
+  // unhandled rejections for now.
+  if (tla_promise.isObject()) {
+    if (!core::EventLoop::run_event_loop(this, 0, &tla_promise)) {
```

Comment:
We're not actually using the `result` parameter in `run_event_loop`, but if we were, this would overwrite the TLA promise.

Instead of taking this suggestion, please feel free to remove the `result` parameter :)
```suggestion
    RootedValue rval(cx);
    if (!core::EventLoop::run_event_loop(this, 0, &rval)) {
```

Reply:
Yeah I wasn't sure what the plan was here. Ok sure, I've replaced this with a pending exception check instead.